src/main/java/ui/components/pickers/MilestonePickerState.java

```diff
@@ -30,78 +34,236 @@ public MilestonePickerState(List<PickerMilestone> milestones, String userInput)
-     * such that the new given milestone is toggled, and the rest are not selected
+     * Changes the selection statuses of all milestones in the allMilestones,
+     * such that the given milestone is toggled, and the rest are not selected
      *
      * @param milestone
      */
     private void toggleMilestone(PickerMilestone milestone) {
-        this.currentMilestonesList.stream()
-                .forEach(listMilestone ->
-                        listMilestone.setSelected(listMilestone.equals(milestone) && !listMilestone.isSelected()));
+        allMilestones.stream()
+                .forEach(listMilestone -> listMilestone.setSelected(listMilestone.equals(milestone)
+                        && !listMilestone.isSelected()));
+    }
+
+    public List<PickerMilestone> getAllMilestones() {
+        return allMilestones;
+    }
+
+    public List<PickerMilestone> getBestMatchingMilestones() {
+        return bestMatchingMilestones;
     }
 
-    public List<PickerMilestone> getCurrentMilestonesList() {
-        return this.currentMilestonesList;
+    /**
+     * Generates the best matching milestones which will then be added to bestMatchingMilestones
+     *
+     * The added milestones are references to the actual elements in allMilestones, and thus should not be
+     * unnecessarily mutated
+     * @param userInput
+     */
+    private void determineBestMatchingMilestones(String userInput) {
+        bestMatchingMilestones.clear();
+        addMatchingMilestones(bestMatchingMilestones, getRemainingBestMatchesToLimit(), allMilestones);
+        if (isBelowLimit(bestMatchingMilestones, BEST_MATCHING_LIMIT)) {
+            addPreviouslyMatchingMilestones(bestMatchingMilestones, getRemainingBestMatchesToLimit(), userInput,
+                                            allMilestones);
+        }
+        if (isBelowLimit(bestMatchingMilestones, BEST_MATCHING_LIMIT)) {
+            addLikelyUnmatchedMilestones(bestMatchingMilestones, getRemainingBestMatchesToLimit(), allMilestones);
+        }
+    }
+
+    private boolean isBelowLimit(List<PickerMilestone> milestones, int limit) {
+        return milestones.size() < limit;
+    }
+
+    private int getRemainingBestMatchesToLimit() {
+        return BEST_MATCHING_LIMIT - bestMatchingMilestones.size();
+    }
+
+    /**
+     * Adds up to {@code limit} number of milestones to {@code bestMatchingMilestones}
+     * These milestones are taken from {@code milestones}, in the same order given,
+     * if they are not already in {@code bestMatchingMilestones}
+     *
+     * This method will do nothing if {@code limit} is not a positive integer
+     * @param bestMatchingMilestones
+     * @param limit
+     * @param milestones
+     */
+    private void addLikelyUnmatchedMilestones(List<PickerMilestone> bestMatchingMilestones, int limit,
+                                              List<PickerMilestone> milestones) {
+        if (limit <= 0) return;
```

Comment:
This method is not supposed to be called with `limit <= 0` right? This kind of 'just in case' coding can hide programmer errors. I'd rather you state the assumption in the method comment and assert it here.